ADAL/src/telemetry/ADTelemetry.m

```diff
@@ -107,39 +107,45 @@ - (NSString*)registerNewRequest
 - (void)startEvent:(NSString*)requestId
          eventName:(NSString*)eventName
 {
-    if ([NSString adIsStringNilOrBlank:requestId] || [NSString adIsStringNilOrBlank:eventName])
+    @synchronized(self)
     {
-        return;
+        if ([NSString adIsStringNilOrBlank:requestId] || [NSString adIsStringNilOrBlank:eventName])
+        {
+            return;
+        }
+        
+        NSDate* currentTime = [NSDate date];
+        [_eventTracking setObject:currentTime
+                           forKey: [self getEventTrackingKey:requestId eventName:eventName]];
     }
-    
-    NSDate* currentTime = [NSDate date];
-    [_eventTracking setObject:currentTime
-                       forKey: [self getEventTrackingKey:requestId eventName:eventName]];
 }
 
 - (void)stopEvent:(NSString*)requestId
             event:(id<ADTelemetryEventInterface>)event
 {
-    NSDate* stopTime = [NSDate date];
-    NSString* eventName = [self getPropertyFromEvent:event propertyName:AD_TELEMETRY_EVENT_NAME];
-    
-    if ([NSString adIsStringNilOrBlank:requestId] || [NSString adIsStringNilOrBlank:eventName] || !event)
-    {
-        return;
-    }
-    
-    NSString* key = [self getEventTrackingKey:requestId eventName:eventName];
-    NSDate* startTime = [_eventTracking objectForKey:key];
-    if (!startTime)
+    @synchronized(self)
     {
-        return;
+        NSDate* stopTime = [NSDate date];
+        NSString* eventName = [self getPropertyFromEvent:event propertyName:AD_TELEMETRY_EVENT_NAME];
+        
+        if ([NSString adIsStringNilOrBlank:requestId] || [NSString adIsStringNilOrBlank:eventName] || !event)
+        {
+            return;
+        }
+        
+        NSString* key = [self getEventTrackingKey:requestId eventName:eventName];
+        NSDate* startTime = [_eventTracking objectForKey:key];
+        if (!startTime)
+        {
+            return;
+        }
+        [event setStartTime:startTime];
+        [event setStopTime:stopTime];
+        [event setResponseTime:[stopTime timeIntervalSinceDate:startTime]];
+        [_eventTracking removeObjectForKey:key];
```

Comment:
Same as above. The only part of this that needs to be protected is the part mutating the dictionary